R/celltype_anno.R

```diff
@@ -131,55 +131,69 @@ perform_goa <- function(gois,
-    gois_in_set <- gois_in_set[gs_names]
-
-    # Build parameter data frame
-    ngois <- length(gois)                   # clustered genes
-    group1 <- lengths(gene_sets)            # the size of gene sets
-    group2 <- length(all_gs)                # total genes in gene sets
-    overlap <- lengths(gois_in_set)         # nr gois in gene_set
-
-    phyper_df <- data.frame(
-        gois_in_set = overlap - 1,          # white balls drawn / gois in gs
-        genes_in_set = group1,              # total white balls / genes in gs
-        genes_universe = group2 - group1,   # total black balls / all genes in gene set - gs
-        ngois = ngois                       # balls drawn / number gois
-    )
-    rownames(phyper_df) <- names(gene_sets)
-
-    # Hypergeometric test
-    pvalues <- apply(phyper_df, 1, function(n) {
-        stats::phyper(n[1], n[2], n[3], n[4], lower.tail = FALSE)
-    })
-
-    # adjust p-values
-    p_adj <- stats::p.adjust(pvalues, method = "BH")
-
-
-    ## gene ratio and background ratio
-    gene_ratio <- apply(data.frame(a = overlap, b = ngois), 1, function(x) {
-        paste(x[1], "/", x[2], sep = "", collapse = "")
-    })
-
-    bg_ratio <- apply(data.frame(a = group1, b = group2), 1, function(x) {
-        paste(x[1], "/", x[2], sep = "", collapse = "")
-    })
-
-    # return results either as data frame or list
-    enrich_res <- data.frame(gene_set = names(gene_sets),
+    if(length(gois_in_set) == 0){
+	print(names(gois_in_set))
+	warning('No cell types can be assigned to the detected marker genes!')
+	enrich_res <- data.frame(gene_set = "None",
+                             pval = NA,
+                             padj = NA,
+                             GeneRatio = NA,
+                             BgRatio = NA,
+                             ngois_in_set = 0,
+                             ngenes_in_set = NA,
+                             ngois = length(gois),
+                             ngenes_in_sets = length(all_gs))
+	return(enrich_res)
+
+	}else{
+
+	# subset gene sets to thos with gois in them
+    	# make sure the two sets are the same order.
+    	gs_names <- sort(unique(names(gois_in_set)))
+    	gene_sets <- gene_sets[gs_names]
+    	gois_in_set <- gois_in_set[gs_names]
+
+    	# Build parameter data frame
+    	ngois <- length(gois)                   # clustered genes
+    	group1 <- lengths(gene_sets)            # the size of gene sets
+    	group2 <- length(all_gs)                # total genes in gene sets
+    	overlap <- lengths(gois_in_set)         # nr gois in gene_set
+
+    	phyper_df <- data.frame(
+        	gois_in_set = overlap - 1,          # white balls drawn / gois in gs
+        	genes_in_set = group1,              # total white balls / genes in gs
+        	genes_universe = group2 - group1,   # total black balls / all genes in gene set - gs
+        	ngois = ngois                       # balls drawn / number gois
+    	)	
+    	rownames(phyper_df) <- names(gene_sets)
```

Comment:
I got an error when running this line with a test data, therfore line141-155 was added to solve it.